File: benchmarks/compute/floyd-warshall/Go/floyd.go
package main

import (
    "fmt"
    "math/rand"
    "os"
    "strconv"
    "time"
)

const (
    SMALL = iota
    STANDARD
    LARGE
)

const DATASET = STANDARD

var N int

func init() {
    switch DATASET {
    case SMALL:
        N = 50
    case LARGE:
        N = 2000
    default:
        N = 500
    }
}

func initMatrix(n int, seed int64) [][]float64 {
    rand.Seed(seed)
    graph := make([][]float64, n)
    for i := range graph {
        graph[i] = make([]float64, n)
        for j := range graph[i] {
            if i == j {
                graph[i][j] = 0.0
            } else {
                graph[i][j] = float64(rand.Intn(100) + 1)
            }
        }
    }
    return graph
}

func printMatrix(n int, graph [][]float64) {
    for i := 0; i < n && i < 8; i++ {
        for j := 0; j < n && j < 8; j++ {
            fmt.Printf("%.2f ", graph[i][j])
        }
        fmt.Println()
    }
}

func floydWarshall(n int, graph [][]float64) {
    for k := 0; k < n; k++ {
        for i := 0; i < n; i++ {
            for j := 0; j < n; j++ {
                if graph[i][j] > graph[i][k]+graph[k][j] {
                    graph[i][j] = graph[i][k] + graph[k][j]
                }
            }
        }
    }
}

func main() {
    seed := int64(42)
    if len(os.Args) > 1 {
        s, _ := strconv.ParseInt(os.Args[1], 10, 64)
        seed = s
    }

    graph := initMatrix(N, seed)

    start := time.Now()
    floydWarshall(N, graph)
    elapsed := time.Since(start)

    fmt.Printf("Finished Floyd-Warshall in %.3f seconds\n", elapsed.Seconds())
    printMatrix(N, graph)
}

File: benchmarks/compute/floyd-warshall/C/floyd.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define SMALL 0
#define STANDARD 1
#define LARGE 2

#if !defined(DATASET)
#define DATASET STANDARD
#endif

#if DATASET == SMALL
#define N 50
#elif DATASET == LARGE
#define N 2000
#else
#define N 500
#endif

void init_matrix(int n, double graph[n][n], unsigned int seed) {
    srand(seed);
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            graph[i][j] = (i == j) ? 0.0 : ((double)(rand() % 100) + 1);
}

void print_matrix(int n, double graph[n][n]) {
    for (int i = 0; i < n && i < 8; i++) {
        for (int j = 0; j < n && j < 8; j++) {
            printf("%0.2f ", graph[i][j]);
        }
        printf("\n");
    }
}

void floyd_warshall(int n, double graph[n][n]) {
    for (int k = 0; k < n; k++)
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                if (graph[i][j] > graph[i][k] + graph[k][j])
                    graph[i][j] = graph[i][k] + graph[k][j];
}

int main(int argc, char** argv) {
    static double graph[N][N];
    unsigned int seed = (argc > 1) ? atoi(argv[1]) : 42;

    init_matrix(N, graph, seed);

    clock_t start = clock();
    floyd_warshall(N, graph);
    clock_t end = clock();

    printf("Finished Floyd-Warshall in %.3f seconds\n",
           (double)(end - start) / CLOCKS_PER_SEC);

    print_matrix(N, graph);
    return 0;
}

